main/src/analytical_solutions/turbulence/density_pdf.cpp

```diff
@@ -0,0 +1,137 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <filesystem>
+#include <fstream>
+#include <numeric>
+
+#include "density_pdf.hpp"
+#include "io/arg_parser.hpp"
+#include "io/factory.hpp"
+#include "util/utils.hpp"
+#include "cstone/primitives/mpi_wrappers.hpp"
+
+using namespace sphexa;
+using T = float;
+
+int main(int argc, char** argv)
+{
+    auto [rank, numRanks] = initMpi();
+    const ArgParser parser(argc, (const char**)argv);
+
+    // The default pdf range was taken in comparison to Federrath et al. 2021, DOI 10.1038/s41550-020-01282-z
+    const std::string inputFile  = parser.get("--file");
+    const size_t      nBins      = parser.get("-n", 50);
+    const int         step       = parser.get("-s", -1);
+    std::string       outputFile = parser.get("-o", std::string("density_pdf.txt"));
+    const std::string sph_type   = parser.get("--sph", std::string("std"));
+    const T           minValue   = parser.get("--min", -8.0);
+    const T           maxValue   = parser.get("--max", 6.0);
+
+    if (!std::filesystem::exists(inputFile))
+    {
+        printf("Please provide a existing file: no file found at %s\n", inputFile.c_str());
+        return exitSuccess();
+    }
+
+    std::vector<T> rho;
+    std::vector<T> bins;
+
+    auto h5reader = fileReaderFactory(false, MPI_COMM_WORLD);
+    h5reader->setStep(inputFile, step, FileMode::collective);
+
+    const size_t localNumParticles  = h5reader->localNumParticles();
+    const size_t globalNumParticles = h5reader->globalNumParticles();
+    if (rank == 0)
+    {
+        printf("Density-PDF: local particles: %lu \t global particles: %lu\n", localNumParticles, globalNumParticles);
+    }
+
+    rho.resize(localNumParticles);
+
+    if (sph_type == "std") { h5reader->readField("rho", rho.data()); }
+    else
+    {
+        std::vector<T> m(localNumParticles);
+        std::vector<T> xm(localNumParticles);
+        h5reader->readField("kx", rho.data());
+        h5reader->readField("xm", xm.data());
+        h5reader->readField("m", m.data());
+#pragma omp for schedule(static)
+        for (size_t i = 0; i < localNumParticles; ++i)
+        {
+            rho[i] = rho[i] * m[i] / xm[i];
+        }
+    }
+    rho.shrink_to_fit();
+    if (rho.size() != localNumParticles)
+    {
+        throw std::runtime_error("rho length doesn't match local count: " + std::to_string(rho.size()) + "\t" +
+                                 std::to_string(localNumParticles));
+    }
+
+    h5reader->closeStep();
+
+    T localTotalDensity = 0.0;
```

Comment:
This variable should be `double` to avoid loss of precision. (`rho` in `float` is fine)